New EEP: Module attributes for documentation (#24)
erlang/eep: eeps/eep-0057.md

    Status: Draft
    Type: Standards Track
    Created: 02-Jun-2021
    Post-History:
****
EEP 57: Module attributes for documentation
----

Abstract
========

This EEP proposes a structured documentation API for Erlang where the documentation is handled as part of the language parser and included directly in the compiled .beam files, as a replacement for EDoc. Python, Elixir, and Clojure are examples of languages that follow this approach of treating documentation as data rather than code comments.

Rationale
=========

The main limitation in EDoc today is that the documentation is kept as code comments. This requires an explicit tool to parse said code comments, which complicates access to the docs by IDEs, from the shell, etc. There have been recent improvements in this area by making EDoc compile to EEP 48 but it still requires an explicit step.

Furthermore, the “code comments” approach is more complex implementation wise, as it requires parsing the source code alongside code comments, parsing the code comments, and so on. Personally, I also enjoy the explicit distinction between documentation and code comments: they have different requirements and different audiences.

> This EEP proposes the addition of two module attributes to Erlang: `-doc` and `-moduledoc`. This EEP also includes an extra section on additional features that can help writing documentation with those attributes more productive, but that's optional. As with EEP 48, this proposal pertains exclusively to API references and their documentation. It doesn't cover guides, tutorials, and other documentation formats.

Author: ```suggestion
This EEP proposes the addition of two module attributes to Erlang: `-doc` and `-moduledoc`. This EEP also includes an extra section on additional features that can make writing documentation more productive, but they are optional and some of them should likely be explored more in depth in other EEPs. Still, including them here can be relevant to provide a long term view and some of challenges found when providing structured documentation.

As with EEP 48, this proposal pertains exclusively to API references and their documentation. It doesn't cover guides, tutorials, and other documentation formats.
```

New module attributes
=====================

This EEP proposes two new attributes: `-doc` and `-moduledoc`. They could be used as follows:

    -module(base64).
    -moduledoc "
    Convenience functions for encoding and decoding from base64.
    ".

    -doc "
    Encodes the given binary to base64.
    ".
    -spec encode(binary()) -> binary().
    encode(Binary) ->
      % ....

    -doc "
    Decodes the given binary from base64.
    ".
    -spec decode(binary()) -> {ok, binary()} | error.
    encode(Binary) ->
      % ....

The new `-moduledoc` attribute can be listed anywhere and it will contain the documentation for the given module. The `-doc` attribute must be listed once before each function and it will contain the documentation for the following function.

> Listing multiple `-doc` attributes for the same function should raise an exception.

Author: ```suggestion
Listing multiple `-doc` attributes for the same function should warn or error accordingly.
```

  1. Use the `-doc` attribute
  2. Introduce distinct `-callbackdoc` and `-typedoc` attributes

Those approaches are mostly equivalent. The second option is more explicit albeit slightly longer to type.

## Metadata

The new module attributes also support documentation metadata by passing a map as argument:

    -module(beam64).
    -moduledoc "
    Convenience functions for encoding and decoding from base64.
    ".
    -moduledoc #{
      author => [<<"The Erlang/OTP team">>],
      license => <<"Apache 2 License">>,
      cross_references => [binary]
    }.

If the `-moduledoc` is called multiple times with a map, the maps will be merged. This comes with the added benefit that shared metadata can be moved to a header file:

    %% prelude.hrl
    -moduledoc #{
      authors => [<<"The Erlang/OTP team">>],
      license => <<"Apache 2 License">>
    }.

which we can then include and augment:

    -module(beam64).
    -include("prelude.hrl").
    -moduledoc "
    Convenience functions for encoding and decoding from base64.
    ".
    -moduledoc #{cross_references => [binary]}.

A list of built-in attributes is available on EEP 48.

## Compilation

> Compiling a module with the `-moduledoc` or `-doc` attributes will automatically generate a Docs chunk into its .beam file, making the documentation directly accessible in the shell. A `-nodocs` flag can be added to `erlc` to skip the generation of the chunk.

Author: ```suggestion
Compiling a module with the `-moduledoc` or `-doc` attributes will automatically generate a Docs chunk into its .beam file, making the documentation directly accessible in the shell. A `-nodocs` flag can be added to `erlc` to skip the generation of the chunk.

Release tools should also prune the Documentation chunk out of .beam files by default. Note this is already done by `beam_lib:strip_release/1` and `beam_lib:strip_files/1`.
```

Note that similar concerns apply to documentation metadata. Attributes such as License, Authors, and so on, are required to be binaries. Since those attributes are known upfront, they can be validated/normalized by the parser/compiler.

With this in mind, it is important to remember that Erlang binaries are latin1 by default. Therefore, if the documentation contains non-latin1 characters, one would have to write:

    -doc <<"Documentation about púnctuation"/utf8>>.

Similarly, the Authors metadata:

    -doc #{
      authors => [<<"José Valim"/utf8>>]
    }.

For this reason, the Build and Packaging Working Group of the Erlang Ecosystem Foundation has also discussed the option of making Unicode binary strings easier to write in Erlang, such as `u"Héllò World"`. While adding such construct is not a direct part of this EEP, the point made in this subsection is that, if we require the documentation or its metadata to be a binary, we may need to provide conveniences for writing UTF-8 binary strings in Erlang.

## Triple-quoted strings

> In the examples above, we have used regular Erlang strings for the documentation. However, they come with a limitation: escaping can become confusing quite fast as soon as the documentation itself contains double quoted strings. One option here is to add triple-quoted strings to Erlang:

Reviewer: I think the triple quote part here is a bit redundant and not obvious. Maybe just put something saying this is an example and keep bringing the triple quote only after this example ?

> In the examples above, we have used regular Erlang strings for the documentation. However, they come with a limitation: escaping can become confusing quite fast as soon as the documentation itself contains double quoted strings. One option here is to add triple-quoted strings to Erlang:
> 
>     -doc "Remove double-quotes (\") from the given string".
>     remove_double_quotes(String) ->
>       ...
> 
> For this reason, we may opt-in to provide triple-quoted strings, which reads better in multi-line format and reduces the need for escaping:

Author: ```suggestion
Another feature often related to structure documentation is triple-quoted strings. To understand why they may be useful, consider the following example that has a double quote in its documentation:

    -doc "Remove double-quotes (\") from the given string".
    remove_double_quotes(String) ->
      ...

For this reason, we may support triple-quoted strings, which reads better in multi-line format and reduces the need for escaping:
```

## Doctests

A direct consequence of making the documentation more structured and accessible is that Erlang can include doctests, which is the ability to run and validate the examples in your documentation. For example, someone could write this:

    -doc """
    Encodes the given binary to base64.

        1> base64:encode("hello").
        <<"aGVsbG8=">>

    """.
    -spec encode(binary()) -> binary().
    encode(Binary) ->
      % ....

And then in your test suite:

    doctests(_Config) ->
      ct_doctest:run(base64).

The doctest attribute will access the documentation entries in the base64 Docs chunk, extract all of the examples, and run them. Of course, there is nothing stopping doctests from being implemented on top of EDoc today, but the fact the documentation is already parsed by the parser/compiler, makes doctests trivial to add. Doctests would benefit from a separate EEP, as there are some extra considerations, as doctesting exceptions, unparseable formats, etc, but it is worth mentioning them given their benefits to users and documentation authors.

EDoc and source code documentation
==================================

## Should documentation be in the source code?

> There is one discussion which is the debate if API Reference documentation should exist in the source code or be placed in a separate file. There are arguments against and in favor of both approaches. If the documentation is alongside the source code, it is easier to keep it up to date. At the same time, it may pollute the source code. This proposal does not advocate that one is better than the other. Rather, because we already have a tool that colocates documentation and source code, namely EDoc, this EEP aims to propose a more integrated variant of said tool.

Author: ```suggestion
There is one discussion which is the debate if API Reference documentation should exist in the source code or be placed in a separate file. There are arguments against and in favor of both approaches. If the documentation is alongside the source code, it is easier to keep it up to date. At the same time, it may pollute the source code. This proposal does not advocate that one is better than the other. Rather, because we already have a tool that colocates documentation and source code, namely EDoc, this EEP aims to propose a more integrated variant of said tool, which will hopefully lead developers to better document their Erlang modules and libraries.
```

## What about EDoc?

If this proposal is to be accepted, what happens with Edoc?

My proposal would be for EDoc to be repurposed with a new backend that translates the existing EDoc documentation to the new module attribute format. Eventually EDoc can be deprecated and phased out.

The HTML rendering part of EDoc can be repurposed to work on the Docs chunk, allowing the Docs chunk to benefit from it. The Build and Packaging Working Group of the Erlang Ecosystem Foundation is also working on getting the ExDoc project ready to generate HTML+EPUB for Erlang projects. The goal is to distribute ExDoc as an escript which will make it really easy to integrate with existing Erlang libraries via a convenient Command Line Interface.

Similarly, if the goal is to eventually phase out EDoc, the Erlang/OTP team needs to discuss what happens with the existing EDoc to XML conversion in `erl_docgen`. Two options exist:

  * Generate EDoc to XML one last time and use the XML as the source from now on

  * Convert EDoc to the documentation style in this proposal and teach erl_docgen how to convert the Docs chunk to XML

Copyright
=========

> This document has been placed in the public domain.

Reviewer 2: This probably should strictly describe license, like CC0, as not all jurisdictions allow voluntary assigning document to public domain.

Author: This is the standard of all other EEPs, so I won't change it, but perhaps this conversation needs to brought up elsewhere? Or maybe it only matters for what is valid in Sweden in this case?

Reviewer 2: @josevalim created #25

Reviewer 3: AFAIK, the templates were copied from PEP, so this phrasing is likely just an old artefact. I agree it should be changed to CC0. In many countries (including most of Europe) saying "this is placed in the public domain" does not have any legal effect. Copyright is automatic as of the Berne convention of 1988.

    Status: Draft
    Type: Standards Track
    Created: 02-Jun-2021
    Post-History:
****
EEP 57: Module attributes for documentation
----

Abstract
========

This EEP proposes a structured documentation API for Erlang where the
documentation is handled as part of the language parser and included
directly in the compiled .beam files, as a replacement for EDoc.
Python, Elixir, and Clojure are examples of languages that follow this
approach of treating documentation as data rather than code comments.

Rationale
=========

The main limitation in EDoc today is that the documentation is kept as
code comments.  This requires an explicit tool to parse said code
comments, which complicates access to the docs by IDEs, from the shell,
etc.  There have been recent improvements in this area by making EDoc
compile to EEP 48 but it still requires an explicit step.

Furthermore, the “code comments” approach is more complex implementation
wise, as it requires parsing the source code alongside code comments,
parsing the code comments, and so on.  Personally, I also enjoy the
explicit distinction between documentation and code comments: they have
different requirements and different audiences.

> This EEP proposes the addition of two module attributes to Erlang: `-doc`
> and `-moduledoc`.  This EEP also includes an extra section on additional
> features that can help writing documentation with those attributes more
> productive, but that's optional. As with EEP 48, this proposal pertains
> exclusively to API references and their documentation.  It doesn't cover
> guides, tutorials, and other documentation formats.

Author: ```suggestion
This EEP proposes the addition of two module attributes to Erlang: `-doc`
and `-moduledoc`.  It also includes an extra section on additional features
that can make aid writing documentation, but they are optional and some
of them should likely be explored more in depth in other EEPs.
Still, they are included to provide a long term view of the features and
challenges related to structured documentation.
          
As with EEP 48, this proposal pertains exclusively to API references and
their documentation.  It doesn't cover guides, tutorials, and other
documentation formats.
```

that can make aid writing documentation, but they are optional and some
of them should likely be explored more in depth in other EEPs.
Still, they are included to provide a long term view of the features and
challenges related to structured documentation.
          
As with EEP 48, this proposal pertains exclusively to API references and
their documentation.  It doesn't cover guides, tutorials, and other
documentation formats.

New module attributes
=====================

This EEP proposes two new attributes: `-doc` and `-moduledoc`.  They could
be used as follows:

    -module(base64).
    -moduledoc "
    Convenience functions for encoding and decoding from base64.
    ".

    -doc "
    Encodes the given binary to base64.
    ".
    -spec encode(binary()) -> binary().
    encode(Binary) ->
      % ....

    -doc "
    Decodes the given binary from base64.
    ".
    -spec decode(binary()) -> {ok, binary()} | error.
    encode(Binary) ->
      % ....

The new `-moduledoc` attribute can be listed anywhere and it will contain
the documentation for the given module.  The `-doc` attribute must be
listed once before each function and it will contain the documentation

> for the following function.

Reviewer 4: Any ordering preferences compared to type specs?

Author: @ferd please expand? I am not sure I follow the question. :)

Reviewer 4: People have a tendency to do something like:

```
%% @doc This function turns the programmer into a much more productive one
-spec levelup(programmer()) -> programmer().
levelup(Programmer) -> Programmer*10.
```

The new literal translation would be:

```
-doc "This function turns the programmer into a much more productive one".
-spec levelup(programmer()) -> programmer().
levelup(Programmer) -> Programmer*10.
```

This will work with the spec above if we read "listed _once before_ each function", meaning there can be one `-doc` attribute before any function and it will get assigned to the one that follows.

However the spec could also force the following if we read it to mean "the doc must be listed once _right before_ each function":

```
-spec levelup(programmer()) -> programmer().
-doc "This function turns the programmer into a much more productive one".
levelup(Programmer) -> Programmer*10.
```

Author: Ah, I see. We don't look at other attributes, so it can be anywhere before a function. I will clarify.

> The new `-moduledoc` attribute can be listed anywhere and it will contain
> the documentation for the given module.  The `-doc` attribute must be
> listed once before each function and it will contain the documentation
> for the following function.

Author: ```suggestion
The new `-moduledoc` attribute can be listed anywhere and it will contain
the documentation for the given module.  The `-doc` attribute must be
listed anywhere before a function and it will contain the documentation
for the following function.  For instance, the example below:

    -doc "Example".
    -spec example() -> ok.
    example() -> ok.

is equivalent to:

    -spec example() -> ok.
    -doc "Example".
    example() -> ok.    
```

> Listing multiple `-doc` attributes for the same function should warn or
> error accordingly.

Author: ```suggestion
Listing multiple `-doc` attributes for the same function should warn or
error accordingly, unless the documentation is being set to hidden.
For example, this is valid:

    -doc "Example".
    -doc hidden.
    example() -> ok.

But this should warn/error:

    -doc "Example".
    -doc "Updated example".
    example() -> ok.

This as well:

    -doc "Example".
    example(one) -> 1;
    -doc "Updated example".
    example(two) -> 2;
```

    at some point
  * Support both prior formats and automatically convert lists to
    binaries at some point - preferably in the parser/compiler,
    to avoid pushing this concern to everyone downstream

Note that similar concerns apply to documentation metadata.
Attributes such as License, Authors, and so on, are required
to be binaries.  Since those attributes are known upfront, they
can be validated/normalized by the parser/compiler.

With this in mind, it is important to remember that Erlang binaries
are latin1 by default. Therefore, if the documentation contains
non-latin1 characters, one would have to write:

    -doc <<"Documentation about púnctuation"/utf8>>.

Similarly, the Authors metadata:

    -doc #{
      authors => [<<"José Valim"/utf8>>]
    }.

For this reason, the Build and Packaging Working Group of the
Erlang Ecosystem Foundation has also discussed the option of making
Unicode binary strings easier to write in Erlang, such as
`u"Héllò World"`.  While adding such construct is not a direct part
of this EEP, the point made in this subsection is that, if we require
the documentation or its metadata to be a binary, we may need to

> provide conveniences for writing UTF-8 binary strings in Erlang.

Reviewer 4: This, to me, would point towards supporting both formats and converting. Erlang strings should represent safe unicode data that can be converted later with `unicode:characters_to_binary(CharData)` and for the time being provides safe unicode with low syntactic overhead.

Author: Yeah, I think I agree with you. What about metadata? Should it be automatically rewritten too?

Reviewer 4: It could make sense. I'd see this as an extension of all Erlang-style things where `CharData` gets flattened down for you when possible. It would in principle not harm anything as long as you specify you expect data to be UTF-8 compatible, meaning raw strings, binaries in the utf8 encoding, or lists of either of these.

> For this reason, the Build and Packaging Working Group of the
> Erlang Ecosystem Foundation has also discussed the option of making
> Unicode binary strings easier to write in Erlang, such as
> `u"Héllò World"`.  While adding such construct is not a direct part
> of this EEP, the point made in this subsection is that, if we require
> the documentation or its metadata to be a binary, we may need to
> provide conveniences for writing UTF-8 binary strings in Erlang.

Author: ```suggestion
For this reason, the Build and Packaging Working Group of the
Erlang Ecosystem Foundation has also discussed the option of making
Unicode binary strings easier to write in Erlang, such as
`u"Héllò World"`.  While adding such construct is not a direct part
of this EEP, the point made in this subsection is that, if we require
the documentation or its metadata to be a binary, we may need to
provide conveniences for writing UTF-8 binary strings in Erlang.
Alternatively, this may also indicate the best option is to support
the documentation to be written as charlists (or `unicode:chardata()`)
and convert them to UTF-8 encoded binaries internally.
```

detailed below, may strongly rely on triple-quoted strings.

## Doctests

A direct consequence of making the documentation more structured
and accessible is that Erlang can include doctests, which is the
ability to run and validate the examples in your documentation.
For example, someone could write this:

    -doc """
    Encodes the given binary to base64.

        1> base64:encode("hello").
        <<"aGVsbG8=">>

    """.
    -spec encode(binary()) -> binary().
    encode(Binary) ->
      % ....

And then in your test suite:

    doctests(_Config) ->
      ct_doctest:run(base64).

The doctest attribute will access the documentation entries in
the base64 Docs chunk, extract all of the examples, and run them.
Of course, there is nothing stopping doctests from being implemented
on top of EDoc today, but the fact the documentation is already
parsed by the parser/compiler, makes doctests trivial to add.
Doctests would benefit from a separate EEP, as there are some extra
considerations, as doctesting exceptions, unparseable formats, etc,
but it is worth mentioning them given their benefits to users and
documentation authors.

> 

Author: ```suggestion

## `warn_missing_doc`

Another optional feature is to support a `-compile(warn_missing_doc).`
that emits a warning whenever an exported function is not documented.

```

challenges related to structured documentation.
          
As with EEP 48, this proposal pertains exclusively to API references and
their documentation.  It doesn't cover guides, tutorials, and other
documentation formats.

New module attributes
=====================

This EEP proposes two new attributes: `-doc` and `-moduledoc`.  They could
be used as follows:

    -module(base64).
    -moduledoc "
    Convenience functions for encoding and decoding from base64.
    ".

    -doc "
    Encodes the given binary to base64.
    ".
    -spec encode(binary()) -> binary().
    encode(Binary) ->
      % ....

    -doc "
    Decodes the given binary from base64.
    ".
    -spec decode(binary()) -> {ok, binary()} | error.
    encode(Binary) ->
      % ....

The new `-moduledoc` attribute can be listed anywhere and it will contain
the documentation for the given module.  The `-doc` attribute must be
listed anywhere before a function and it will contain the documentation
for the following function.  For instance, the example below:

    -doc "Example".
    -spec example() -> ok.
    example() -> ok.

is equivalent to:

    -spec example() -> ok.
    -doc "Example".
    example() -> ok.    

> Listing multiple `-doc` attributes for the same function should warn or
> error accordingly, unless the documentation is being set to hidden.
> For example, this is valid:

Reviewer 2: This probably should be:

```suggestion
Listing multiple `-doc` attributes with string for the same function
should warn orerror accordingly, unless the documentation is being
set to hidden.
For example, this is valid:
```

To not conflict with the metadata section.

Reviewer 4: `orerror` sounds like a typo though

Reviewer 2: @ferd fixed.
